tests/test_pythonpackage.py

```diff
@@ -0,0 +1,133 @@
+
+import os
+import shutil
+import sys
+import subprocess
+import tempfile
+
+from pythonforandroid.pythonpackage import (
+    get_package_dependencies,
+    get_dep_names_of_package,
+    get_package_name,
+    _get_system_python_executable,
+)
+
+
+def test_get_package_name():
+    assert get_package_name("PiLLOw") == "Pillow"
+
+
+def test_get_package_dependencies():
+    # Check that jinja2 is returned as direct dep:
+    assert(len([dep for dep in get_package_dependencies("python-for-android")
+                if "jinja2" in dep]) > 0)
+    # Check that MarkupSafe is returned as indirect dep of jinja2:
+    assert [
+        dep for dep in get_package_dependencies(
+            "python-for-android", recursive=True
+        )
+        if "MarkupSafe" in dep
+    ]
+    # Check that colorama is returned without the install condition when
+    # just getting the names:
+    assert "colorama" in get_dep_names_of_package("python-for-android")
+
+
+def test__get_system_python_executable():
+    # This is kind of absurd that we need this to run the test at all,
+    # but we can't test this inside tox's virtualenv:
+    pybin = _get_system_python_executable()
+
+    # The python binary needs to match our major version to be correct:
+    pyversion = subprocess.check_output([
+        pybin, "-c", "import sys; print(sys.version)"
+    ], stderr=subprocess.STDOUT)
+    try:
+        pyversion = pyversion.decode("utf-8", "replace")
+    except AttributeError:
```

Comment:
I'm curious about that line?

Reply:
the decode? just makes sure it's a unicode string, `subprocess.check_output()` returns bytes sometimes (I have to admit I can't say off the top of my head under which)

Reply:
No about the try/catch in a unit test. In general try catching in unit tests is very fishy

Reply:
@AndreMiras as always I'm happy to remove it if you think I should, *but* I am pretty sure this is the agreed upon "pythonic" way to decode a unicode bytearray everywhere, so I'm really not sure what the issue is...?

I do also use try catch in another place to throw the error again with a better description, and I really don't think there is any issue with it.

(or the other place where I explained in a comment after checking for "pep517" that the test cannot reasonably run since that is not available in the host python)

I agree with the general notion if there isn't a really good reason for it, but in this particular case I'm not sure I can understand your dislike :question: nevertheless, you want me to get it out then I'll remove it and use `type()` or whatever

Reply:
TL;DR: not a hill I want to die on, but I think this is the best way to write it with my limited knowledge

Reply:
Well if you can tell me on what circonstances it would raise an `AttributeError` then maybe I would probably better understand why we need a try/catch here in this unit test :thinking: 
Because I didn't find any obvious reason so far:
```sh
python2 -c 'import subprocess; print(subprocess.check_output(["uname"]).decode("utf-8"))'
Linux

python3 -c 'import subprocess; print(subprocess.check_output(["uname"]).decode("utf-8"))'
Linux
```
In general try/catching in unit test is a code smell, of course there are exception, but they probably deserve a code comment to be justified.
So if this `subprocess.check_output()` call sometimes returns something unexpected I would definitely love to have a clear comment explaining why and under what conditions.
The best way to find out is maybe to drop this try/catch and see if it's actually catching something, if it does then we can document it. If not, great we simplified the code a bit further

Reply:
> In general try/catching in unit test is a code smell

Have to disagree here. How is`try .. catch` not *always* code smell if it's not for a purpose or too wide, and how is it ever smelly if it's narrow with a clear purpose?

The main problem here seems to be that you missed why I use it here, so let me explain:

> Well if you can tell me on what circonstances it would raise an AttributeError

https://docs.python.org/3/library/subprocess.html#subprocess.check_output

As you can read here, it can return bytes, *or* an str. depends on arguments. I would consider it more of a code smell to hardcode the decode as you do, because people tend to adjust parameters and it will blow up then - unless you write it as I did. I suppose you could argue let it blow up if people change the arguments and fix it, but why waste the future maintainers time with that if it can be written in a way that doesn't have that problem?

> The best way to find out is maybe to drop this try/catch and see if it's actually catching something,

Disagree very much. That'd be guessing around a bog standard function which is `subprocess.check_output` of which the behavior of it is well understood and documented. (again see doc link above)

I have to admit I had to look this up and did as you brought it up, because I only vaguely remembered that it could maybe return both types - but it really can, so I think the code makes perfect sense as-is

> but they probably deserve a code comment to be justified.

I disagree again, with this instance specifically. 100% standard approach of "make this a unicode string if it's not already one". I would have to explain this every time I use `subprocess.check_output`, that doesn't sound like a useful thing to do if we want to have a good signal to noise ratio in the code comments

> Again this is in the context of unit testing [...] But specially have one in a unit test looks weird to me

Another disagree :grimacing: to me that's not a good approach/not a useful criteria. Either it has a use or not, and here it has one

These are all just my personal opinions of course. As usual I'll go with what the team decides, and I can be wrong about everything (and I often am), and this can for sure also be written without `try...catch`. However, this is quite a regular Pythonic approach *as far as I'm aware*, by using the object with the intended behavior (decode) and catch the unusual cases (already string, can't decode -> AttributeError) instead of using `type()` hardcoding any of the type checks.

But if you have a better code suggestion let me know! :+1: I think your `subprocess(...).decode` isn't much better though. I suppose you could say if someone changes the arguments and it blows up they'll notice, but I don't see why it shouldn't be written in the first place such that it doesn't matter...?

Reply:
> As you can read here, it can return bytes, or an str. depends on arguments. I would consider it more of a code smell to hardcode the decode as you do, because people tend to adjust parameters and it will blow up then - unless you write it as I did. I suppose you could argue let it blow up if people change the arguments and fix it, but why waste the future maintainers time with that if it can be written in a way that doesn't have that problem?

Everything lies in this statement... You say "because people tend to adjust parameters and it will blow up then" but that's the point, it's an unit test, so you know its entry upfront, this is what you communicate with your test. You say "I know that with input X it gives Y". If your test says "I'm not sure what X is and also not sure what Y would give" then what you're actually testing is not clear...

A good way is to simply refer to some "official" implementation, test or documentation. Here's the cpython one...
https://github.com/python/cpython/blob/v3.7.2/Lib/test/test_subprocess.py#L149

Reply:
Another reference, Django, widely used framework. All the references to subprocess in tests:

- https://github.com/django/django/blob/2.1.7/tests/i18n/test_compilation.py
- https://github.com/django/django/blob/2.1.7/tests/runtests.py
- https://github.com/django/django/blob/2.1.7/tests/admin_scripts/tests.py
- https://github.com/django/django/blob/2.1.7/tests/dbshell/test_postgresql_psycopg2.py

None of them do try/catch in here. Again try/catch or code branching is not forbidden, but often a code smell and should always be justified.
In this case I would just drop it

Reply:
Sorry, I never posted a summary of our followup chat discussion:

I came to the conclusion that while I personally think the bytes/str difference shouldn't be important, it is still *there* and to people who know exactly what it returns, the try/catch conditional decode could be highly confusing (in a way like it was to @AndreMiras ). So while I think I had good motivations, they're probably not worth it in terms of the potential confusion of readers and I've come to agree with @AndreMiras that it should be removed